packages/langium/src/documents/document-builder.ts

```diff
@@ -33,13 +32,12 @@ export class DefaultDocumentBuilder implements DocumentBuilder {
         const parseResult = this.parser.parse(document);
         document.parseResult = parseResult;
         this.process(document);
+        const diagnostics = this.documentValidator.validateDocument(document);
+        document.diagnostics = diagnostics;
 
         if (this.connection) {
-            const diagnostics = this.documentValidator.validateDocument(document);
             // Send the computed diagnostics to VS Code.
             this.connection.sendDiagnostics({ uri: document.uri, diagnostics });
-        } else {
-            resolveAllReferences(parseResult.value);
```

Comment:
Attaching the diagnostics to the document might unnecessarily increase memory usage. We could add an optional callback parameter `(diagnostics: Diagnostic[]) => void` to the `build` method instead.